src/tpm2.rs

```diff
@@ -0,0 +1,957 @@
+    }
+
+    fn pcr_mask_to_list(
+        &self,
+        mask: String,
+        hash_alg: String,
+    ) -> Result<String, KeylimeTpmError> {
+        let mut pcr_list = Vec::new();
+        let mut ima_appended = String::new();
+
+        for pcr in 0..24 {
+            let check_result = self.check_mask(&mask, pcr)?;
+            if check_result {
+                if hash_alg == "SHA1" && pcr == 10 {
+                    ima_appended.push_str(format!("+sha1:{}", pcr).as_str());
+                } else {
+                    pcr_list.push(pcr.to_string());
+                }
+            }
+        }
+        let mut result: String = pcr_list.as_slice().join(",");
+        result.push_str(&ima_appended);
+        Ok(result)
+    }
+
+    fn check_mask(
+        &self,
+        ima_mask: &str,
+        ima_pcr: i32,
+    ) -> Result<bool, KeylimeTpmError> {
+        if ima_mask.is_empty() {
+            return Ok(false);
+        }
+        let ima_mask_int: i32 = ima_mask.parse()?;
+        Ok((1 << ima_pcr) & ima_mask_int != 0)
+    }
+
+    fn temp_file_get_path<'a>(
+        &self,
+        ref temp_file: &'a NamedTempFile,
+    ) -> Result<&'a str, KeylimeTpmError> {
+        temp_file.path().to_str().ok_or_else(|| {
+            KeylimeTpmError::new_tpm_rust_error(
+                "Can't retrieve temp file path.",
+            )
+        })
+    }
+
+    fn tpm_get_manufacturer(&self) -> Result<String, KeylimeTpmError> {
+        self.run("tpm2_getcap -c properties-fixed".to_string(), None)
+            .and_then(|(ret_out, _)| {
+                serde_json::from_str(&ret_out)
+                    .map_err(|e| e.into())
+                    .and_then(|ret_out_map: Value| {
+                        match ret_out_map["TPM_PT_VENDOR_STRING_1"]["value"]
+                            .as_str()
+                        {
+                            None => Err(KeylimeTpmError::new_tpm_rust_error(
+                                "Error: manufacturer not found.",
+                            )),
+                            Some(content) => Ok(content.to_string()),
+                        }
+                    })
+            })
+    }
+
+    fn is_software_tpm(&self) -> bool {
+        match self.tpm_get_manufacturer() {
+            Ok(data) => data == "SW",
+            Err(_) => false,
+        }
+    }
+
+    fn is_vtpm(&self) -> bool {
+        return false;
+    }
+
+    fn is_deep_quote(&self, quote: String) -> bool {
+        match &quote[0..1] {
+            "d" => true,
+            "r" => false,
+            _ => {
+                warn!("Invalid quote type {}.", quote);
+                false
+            }
+        }
+    }
+
+    fn random_password(
+        &self,
+        length: usize,
+    ) -> Result<String, KeylimeTpmError> {
+        let rand_byte = crypto::generate_random_byte(&length)?;
+        let alphabet: Vec<char> =
+            "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGIJKLMNOPQRSTUVWXYZ"
```

Comment:
Add 3 more characters to this?  (Or document why not.)

Reply:
What other characters are you suggesting to add to the string?

Reply:
Right now you have upper and lowercase alphabet, plus numbers - that's 62, or would be if you hadn't left one out.  (Uppercase 'H'.)  Ideally, it would be 64.  base64 typically uses + and /.

This also allows an alternate, cleaner approach: generate random bytes of the appropriate length, then base64 encode them.